ykrt/pass_finder/try_passes.py

```diff
@@ -66,53 +66,79 @@ def log(logf, s):
 def executable_exists(cmd):
     return shutil.which(cmd) is not None
 
-def get_all_passes(is_prelink):
-    cmd = get_opt_cmd(is_prelink) 
-    p = Popen(cmd, shell=True, stdout=PIPE, close_fds=True)
-    sout, serr = p.communicate() 
-    assert(p.returncode == 0)
+def split_passes(passes_string):
+    parts = []           
+    temp_part = []      
+    paren_stack = []     
+    angle_stack = []    
+
+    for c in passes_string:
+        if c == '(':
+            paren_stack.append(c) 
+        elif c == ')':
+            paren_stack.pop() 
+        
+        if c == '<':
+            angle_stack.append(c) 
+        elif c == '>':
+            angle_stack.pop() 
+
+        if c == ',' and not paren_stack and not angle_stack:
+            part = ''.join(temp_part).strip()
+            if part: 
+                parts.append(part)
+            temp_part = []
+        else:
+            temp_part.append(c)
+ 
+    part = ''.join(temp_part).strip()
+    if part:  
+        parts.append(part)
 
-    sout = sout.decode()
-    pass_descrs = [x.strip() for x in sout.strip().split("\n")]
+    return parts
 
+def get_all_passes(is_prelink):
+    cmd = get_opt_cmd(is_prelink)
+    sout = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+    sout = sout.stdout.decode('utf-8')
+    pass_descrs = split_passes(sout)
     passes = []
     seen = set()
-     
+
     for descr in pass_descrs:
         if descr not in seen:
             seen.add(descr)
-            passes.append(Pass(descr))
-    
-
+            if descr != "BitcodeWriterPass":
+                passes.append(Pass(descr))
+      
     print(f"Found {len(passes)} passes")    
     return passes
 
-def test_pipeline(logf, pl):
+def test_pipeline(logf, pl, cwd):
     sys.stdout.write(str(pl) + "...")
     sys.stdout.flush()
 
-    log(logf, "\n\n" + str(pl) + "\n")  
+    log(logf, "\n\n" + str(pl) + "\n")
 
     # Make sure we don't run empty strings in pipeline.
     assert (len(pl.pre_link) != 0 or len(pl.link_time) != 0), "Both prelink and postlink passes cannot be empty!!!"
 
-    env = os.environ
+    env = os.environ.copy()  # Create a copy of the environment
     env["PRELINK_PASSES"] = ",".join([p.name for p in pl.pre_link])
-    env["LINKTIME_PASSES"] = ",".join([p.name for p in pl.link_time])
-
-    p = Popen("try_repeat 10 sh run_tests.sh 2>&1", cwd=CWD, shell=True,
-              stdout=PIPE, close_fds=True, env=env)
+    env["POSTLINK_PASSES"] = ",".join([p.name for p in pl.link_time])
     
-    sout, _ =  p.communicate()
-    log(logf, sout.decode())
+    ret, time = cargo_run.run_test(cwd, env=env)
 
-    if p.returncode == 0:
+    print(f"\033[92m return code for cargo_run: {ret} \033[0m")
```

Comment:
Same comment as earlier.

Reply:
this is already handled, when returncode `ret` is non zero `test_pipeline` returns `False, None`

Reply:
So this `print` can't ever fire?

Reply:
I have removed this `print` one of the commits I pushed yesterday. The returncode is being check in the next line with `if ret == 0:`